src/ConvertToVersionRange.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class ConvertToVersionRange : Task
+{
+    [Required]
+    public ITaskItem[] References { get; set; } = [];
+
+    [Required]
+    public string VersionProperty { get; set; } = string.Empty;
+
+    [Output]
+    public ITaskItem[] ReferencesWithVersionRanges { get; private set; } = [];
+
+    public override bool Execute()
+    {
+        var success = true;
+
+        foreach (var reference in References)
+        {
+            var automaticVersionRange = reference.GetMetadata("AutomaticVersionRange");
+
+            if (automaticVersionRange.Equals("false", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            var privateAssets = reference.GetMetadata("PrivateAssets");
+
+            if (privateAssets.Equals("All", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            var version = reference.GetMetadata(VersionProperty);
+            var match = Regex.Match(version, @"^\d+");
```

Comment:
Given how the code is being included in MSBuild, can it use the Regex source generators?

Reply:
Nope

Reply:
Assuming it does only get compiled once during a build, it would probably still make sense to make it a:

```
static readonly Regex VersionNumberRegex = new Regex(@"^\d+", RegexOptions.Compiled);
```

Reply:
Given how simple the regex is, I'm not sure it's worth bothering with, but I'll give it a try and see how it impacts compilation time for the task.

Reply:
Also, it can compile more than once, depending on how many projects start building in parallel that need the task when it hasn't been compiled and cached.